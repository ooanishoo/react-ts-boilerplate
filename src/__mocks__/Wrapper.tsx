import React from 'react'
import { BrowserRouter as Router } from 'react-router-dom'

/* a wrapper Router component that takes any React component and renders as it's child component */
type WrapperProps = {
    component?: React.ReactNode
}
export const Wrapper: React.FC<WrapperProps> = ({ component }: WrapperProps) => <Router>{component}</Router>
