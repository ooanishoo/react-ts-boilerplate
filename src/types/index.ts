/** All types goes here*/
export type Todo = {
    userId: number
    id: number
    title: string
    completed: boolean
}
