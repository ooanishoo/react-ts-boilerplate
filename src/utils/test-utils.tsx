import React, { ComponentType, ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { Provider } from 'react-redux'
import { BrowserRouter as Router } from 'react-router-dom'
import store from '../redux/store'

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
    return (
        <Provider store={store}>
            <Router>{children}</Router>
        </Provider>
    )
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
const customRender = (ui: ReactElement, { route = '/' } = {}, options?: Omit<RenderOptions, 'queries'>) => {
    // Wrap dispatch in a mock so it can be spied on.
    store.dispatch = jest.fn(store.dispatch)

    window.history.pushState({}, 'Test page', route)

    const newRender = render(ui, { wrapper: AllTheProviders as ComponentType, ...options })
    return { store, ...newRender }
}

// re-export everything if you want to use test-utils file to import everything
// export * from '@testing-library/react'

// override render method
export { customRender as render }
