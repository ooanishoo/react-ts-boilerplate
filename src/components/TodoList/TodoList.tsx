import { map } from 'ramda'
import React, { useEffect, useRef, useState } from 'react'
import { Link } from 'react-router-dom'
import { getTodos } from '../../api'
import { Todo } from '../../types'

const TodoList: React.FC = () => {
    const [todos, setTodos] = useState<Todo[]>([])
    const [isLoading, setIsLoading] = useState<boolean>(true)
    const [error, setError] = useState('')
    const isMounted = useRef<boolean>(true)

    useEffect(() => {
        return () => {
            // called when the component is going to unmount
            isMounted.current = false
        }
    }, [])

    useEffect(() => {
        setTimeout(() => {
            if (isMounted.current) {
                getTodos()
                    .then((todos) => setTodos(todos))
                    .catch((err) => setError(err?.message))
                    .finally(() => setIsLoading(false))
            }
        }, 900)
    }, [])

    if (isLoading) return <div data-testid="loading-spinner">Loading...</div>
    if (error) return <div data-testid="error">{error}</div>
    return (
        <div data-testid="todo-container">
            <h1>This is a Todo List</h1>
            <ul data-testid="todo-list">
                {map((todo: Todo) => (
                    <li data-testid="todo-item" key={todo.id}>
                        <Link to={{ pathname: `/todos/${todo.id}`, state: { todo } }}>
                            <input type="checkbox" checked={todo?.completed} readOnly />
                            {todo.title}
                        </Link>
                    </li>
                ))(todos)}
            </ul>
        </div>
    )
}

export default TodoList
