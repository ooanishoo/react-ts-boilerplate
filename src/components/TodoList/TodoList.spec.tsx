import { cleanup, render, waitFor, within } from '@testing-library/react'
import axios, { AxiosResponse } from 'axios'
import React from 'react'
import { act } from 'react-dom/test-utils'
import { Todo } from '../../types'
import { Wrapper } from '../../__mocks__/Wrapper'
import TodoList from './TodoList'

afterEach(cleanup)

jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

const todos: Todo[] = [
    {
        userId: 1,
        id: 1,
        title: 'todo-test-1',
        completed: false
    },
    {
        userId: 2,
        id: 2,
        title: 'todo-test-2',
        completed: true
    }
]

describe('<TodoList/>', () => {
    test('should show loading spinner on component mount', async () => {
        const mockedResponse: AxiosResponse = {
            data: todos,
            status: 200,
            statusText: 'OK',
            headers: {},
            config: {}
        }
        mockedAxios.get.mockResolvedValueOnce(mockedResponse)

        await act(async () => {
            const { getByTestId } = render(<Wrapper component={<TodoList />} />)
            expect(getByTestId('loading-spinner')).toHaveTextContent('Loading...')
            await waitFor(() => expect(getByTestId('loading-spinner')).toBeInTheDocument())
            expect(mockedAxios.get).not.toHaveBeenCalled()
        })
    })

    test('should show error message if the api throws an error ', async () => {
        const mockedResponse: AxiosResponse = {
            data: null,
            status: 404,
            statusText: 'OK',
            headers: {},
            config: {}
        }
        mockedAxios.get.mockResolvedValueOnce(mockedResponse)
        await act(async () => {
            const { getByTestId } = render(<Wrapper component={<TodoList />} />)
            await waitFor(() => expect(getByTestId('error')).toBeInTheDocument())
            expect(getByTestId('error')).toHaveTextContent('Something went wrong')
            expect(mockedAxios.get).toHaveBeenCalled()
        })
    })

    test('should fetch todos and display data', async () => {
        const mockedResponse: AxiosResponse = {
            data: todos,
            status: 200,
            statusText: 'OK',
            headers: {},
            config: {}
        }
        mockedAxios.get.mockResolvedValueOnce(mockedResponse)

        await act(async () => {
            const { getByTestId } = render(<Wrapper component={<TodoList />} />)
            expect(getByTestId('loading-spinner')).toHaveTextContent('Loading...')
            expect(mockedAxios.get).not.toHaveBeenCalled()

            const todoContainer = await waitFor(() => getByTestId('todo-container'))
            expect(todoContainer).toBeInTheDocument()
            const todoList = await waitFor(() => getByTestId('todo-list'))
            expect(todoList).toBeInTheDocument()
            const todo = within(todoList).getAllByTestId('todo-item')
            expect(todo.length).toEqual(2)
            expect(mockedAxios.get).toHaveBeenCalledTimes(1)
        })
    })
})
