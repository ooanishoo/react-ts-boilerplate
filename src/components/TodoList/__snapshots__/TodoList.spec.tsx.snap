// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList/> should render without crashing 1`] = `
<div>
  <div
    data-testid="todo-list"
  >
    This is a Todo List
  </div>
</div>
`;
